python/tk_multi_publish2/api/graph.py

```diff
@@ -0,0 +1,591 @@
+        for item_id in ids_to_del:
+            del self._child_lookup[item_id]
+
+        # ---- TASKS
+
+        # reset the tasks based on the items we're keeping. go through the items
+        # we're keeping (already populated in self._items) and make a list of
+        # all the tasks assigned to them (the ones we're keeping)
+        tasks_to_keep = []
+        for item in self._items:
+            if item.id in self._tasks_by_item:
+                for task_id in self._tasks_by_item[item.id]:
+                    tasks_to_keep.append(self._tasks_by_id[task_id])
+
+        # reset the list of tasks now that we know what we're keeping
+        self._tasks = tasks_to_keep
+
+        # rebuild the task by id lookup. like items, simply build the
+        self._tasks_by_id = {}
+        for task in self._tasks:
+            self._tasks_by_id[task.id] = task
+
+        # clean up the tasks by item lookup
+        ids_to_del = []
+        for item_id in self._tasks_by_item:
+            if item_id not in self._items_by_id:
+                ids_to_del.append(item_id)
+        for item_id in ids_to_del:
+            del self._tasks_by_item[item_id]
+
+        # clean up the item by task lookup
+        ids_to_del = []
+        for task_id in self._items_by_task:
+            if task_id not in self._tasks_by_id:
+                ids_to_del.append(task_id)
+        for task_id in ids_to_del:
+            del self._items_by_task[task_id]
+
+        # ---- PLUGINS
+
+        # determine the plugins used by the remaining tasks
+        plugins_to_keep = set()
+        for task in self._tasks:
+            plugins_to_keep.add(task.plugin)
+
+        # reset the plugin lookup
+        self._plugins = {}
+        for plugin in list(plugins_to_keep):
+            self._plugins[plugin.id] = plugin
+
+    def _traverse_graph(self, parent_item):
+        """Depth first traversal of the supplied item.
+
+        :param parent_item: The item to begin traversing
+        """
+
+        for item in self.get_children(parent_item):
+            yield item
+            for child_item in self._traverse_graph(item):
+                yield child_item
+
+
+class _PublishGraphUnpicklerFactory(object):
+    """
+    This is a factory class for returning an instance of Unpickler to handle
+    serialized publish graphs. The Unpickler instance returned depends on
+    whether ``pickle`` or ``cPickle`` is being used.
+    """
+
+    MAPPED_MODULES = {}
+
+    @classmethod
+    def unpickle_find_class(cls, module_name, class_name):
+        """
+        This is the custom unpickle logic for handling toolkit namespaced
+        imports. It handles finding the imported module/class that matches the
+        serialized one.
+        """
+
+        # given a namespaced module/classname, return the same module in the new
+        # tk namespace. example module name:
+        #     tkimp4b0b5e4c49974e21813a21928b101a9f.tk_multi_publish2.base_hooks
```

Comment:
Maybe it's just because I'm not super familiar with pickle, but how would the unpickler deal with a class rename or a class broken up in two?

At least with JSON we control the entire serialization/deserialization process. Also, JSON allows offline editing of the file, since it isn't tied to any Python logic. That might be useful for TDs who want to patch something on a series of file.